appstore/validator_test.go

```diff
@@ -269,58 +271,40 @@ func TestVerifySandboxReceiptFailure(t *testing.T) {
 	}
 }
 
-func badPayload() http.Handler {
-	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		if "POST" == r.Method {
-			w.Header().Set("Content-Type", "application/json")
-			w.Write([]byte(`{"status": 21002}`))
-			return
-		} else {
-			w.Write([]byte(`unsupported request`))
-		}
+func TestCannotReadBody(t *testing.T) {
+	client := New()
+	testResponse := http.Response{Body: ioutil.NopCloser(errReader(0))}
 
-		w.WriteHeader(http.StatusBadRequest)
-	})
+	if client.parseResponse(&testResponse, IAPResponse{}, http.Client{}, IAPRequest{}) == nil {
+		t.Errorf("expected redirectToSandbox to fail to read the body")
+	}
 }
 
-func redirectToSandbox() http.Handler {
-	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		if "POST" == r.Method {
-			w.Header().Set("Content-Type", "application/json")
-			w.Write([]byte(`{"status": 21007}`))
-			return
-		} else {
-			w.Write([]byte(`unsupported request`))
-		}
+func TestCannotUnmarshalBody(t *testing.T) {
+	client := New()
+	testResponse := http.Response{Body: ioutil.NopCloser(strings.NewReader(`{"status": true}`))}
 
-		w.WriteHeader(http.StatusOK)
-	})
+	if client.parseResponse(&testResponse, StatusResponse{}, http.Client{}, IAPRequest{}) == nil {
+		t.Errorf("expected redirectToSandbox to fail to unmarshal the data")
+	}
 }
 
-func sandboxSuccess() http.Handler {
-	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		if "POST" == r.Method {
-			w.Header().Set("Content-Type", "application/json")
-			w.Write([]byte(`{"status": 0}`))
-			return
-		} else {
-			w.Write([]byte(`unsupported request`))
-		}
+type errReader int
 
-		w.WriteHeader(http.StatusOK)
-	})
+func (errReader) Read(p []byte) (n int, err error) {
+	return 0, errors.New("test error")
 }
 
-func invalidResponse() http.Handler {
+func serverWithResponse(response string) http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 		if "POST" == r.Method {
 			w.Header().Set("Content-Type", "application/json")
-			w.Write([]byte(`qwerty!@#$%^`))
+			w.Write([]byte(response))
 			return
 		} else {
 			w.Write([]byte(`unsupported request`))
 		}
 
-		w.WriteHeader(http.StatusOK)
+		w.WriteHeader(http.StatusBadRequest)
```

Comment:
Is there any test written for the case when the response is other than 200? Basically I want to make sure this module handles the case when there is a network issue between Apple server.

Reply:
Oops! That was an oversight when I was refactoring the tests to make them shorter. Nice catch! 

Corrected